tester/messagetracker.go

```diff
@@ -0,0 +1,95 @@
+package tester
+
+// MessageTracker tracks message offsets for each topic for convenient
+// 'expect message x to be in topic y' in unit tests
+type MessageTracker struct {
+	t                T
+	nextMessageIndex map[string]int
+	tester           *Tester
+}
+
+func newMessageTracker(tester *Tester, t T) *MessageTracker {
+	return &MessageTracker{
+		t:                t,
+		nextMessageIndex: make(map[string]int),
+		tester:           tester,
+	}
+}
+
+// NextMessage returns the next message from the topic since the last time this
+// function was called (or MoveToEnd)
+func (mt *MessageTracker) NextMessage(topic string) (string, interface{}, bool) {
+
+	key, msgRaw, hasNext := mt.NextMessageRaw(topic)
+
+	if !hasNext {
+		return key, msgRaw, hasNext
+	}
+
+	decoded, err := mt.tester.codecForTopic(topic).Decode(msgRaw)
+	if err != nil {
+		mt.t.Fatalf("Error decoding message: %v", err)
+	}
+	return key, decoded, true
+}
+
+// NextMessageRaw returns the next message in passed topic
+func (mt *MessageTracker) NextMessageRaw(topic string) (string, []byte, bool) {
+	q := mt.tester.queueForTopic(topic)
+	if mt.nextMessageIndex[topic] >= q.size() {
+		return "", nil, false
+	}
+	msg := q.message(mt.nextMessageIndex[topic])
+
+	mt.nextMessageIndex[topic]++
+	return msg.key, msg.value, true
+}
+
+// ExpectEmpty ensures the topic does not contain more messages
+func (mt *MessageTracker) ExpectEmpty(topic string) {
```

Comment:
This is not necessary since once can call `Next()` until it returns false

Reply:
actually I'd leave the message, as it does more than just checking we're at the end. Also having a shortcut for the alternative loop that we would have to replae in all projects would be helpful.
but I'd rename it to `ExpectAtEnd` to have it consistend with `MoveToEnd`.
What do you think?

Reply:
What about replacing `ExpectAtEnd()` with something like `mt.Offset() == mt.HWM()` ?

and  `MoveToEnd()` replacing with `mt.Seek(mt.HWM())`?

Reply:
agreed